### PR TITLE
Discussion on rounding

### PR DESCRIPTION
**Pull request by fedgiac**
A draft of my considerations on the likelihood of rounding issues when transferring tokens. So far I didn't spot any case in which rounding might cause a frozen auction.

Inspired by #28.

### CODE REVIEW

File: rounding.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+or at a higher price [12].
+Assuming the initial auction price is used [11], then it means that the auction
+has been settled at the price determined in case [16] (out of the four possible
+cases [13], [14], [15], [16] of `settleAuction`). Then:
+```
+[16]: volumeClearingPriceOrder = currentBidSum * fullAuctionedAmount / minAuctionedBuyAmount
+[11]: auctioningTokenAmount = fullAuctionedAmount - volumeClearingPriceOrder
+      biddingTokenAmount = volumeClearingPriceOrder * minAuctionedBuyAmount / fullAuctionedAmount
+ =>   biddingTokenAmount <= currentBidSum
+``` //[22]
```
Comment (josojo) [inline rounding.md]: I can not find [22] above.

Comment (josojo) [reply rounding.md]: Ah this is a reference to a new equation. got it. Not sure how to make it better in md...

Comment (fedgiac) [reply rounding.md]: I'm currently drafting this file, so editorial quality is lacking. In principle I could use markdown references, but I'll only do that after I'm finished.

File: rounding.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+ -    currentBidSum = sum(orderSellAmount) <= minAuctionedBuyAmount
+ =>   sum(out_per_order) =  sum(orderSellAmount * sellAmount / buyAmount)
+                         <= sum(orderSellAmount) * sellAmount / buyAmount
+                         =  currentBidSum * sellAmount / buyAmount
+``` \\[21]
+Next, we analyze the amount transferred when paying out fees. The fee retrieval
+triggers case [19], causing the following amount of tokens to be sent out:
+```
+[20]: feeAmount = sellAmount * feeNumerator / FEE_DENOMINATOR
+[19]: auctioningTokenAmount = sellAmount - volumeClearingPriceOrder
```
Comment (josojo) [inline rounding.md:101]:
```
auctioningTokenAmount = sellAmount - volumeClearingPriceOrder
```
Is this needed? I tthink everything makes sense even without this line.

File: rounding.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+always triggered. This means that no overflows happens for bid tokens.
+Next, we consider auction tokens. Funds flow out when a user claims order funds
+[17], the fees are paid out with `claimFees` and the unsold auction tokens are
+sent back to the auctioneer ([11], discussed before).
+: case [17] determines that the following amount
+of auctioned tokens is withdrawn for each order by a user:
+```
+[17]: out_auction_token_per_order = sellAmount * priceNumerator / priceDenominator
+``` 
+We argue that `sum(out_per_order) <= currentBidSum * sellAmount / buyAmount`.
```
Comment (josojo) [inline rounding.md:84]: Is there some inconsistency? Frist using the one and then later the other notation?
```
sellAmount / buyAmount
<=>
fullAuctionedAmount / minAuctionedBuyAmount 
```

File: rounding.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+```
+We saw all ins/outs un the case where the auction settles with the initial
+auction price. We can add up all transfers involving the auction token to see
+that no more auction tokens as available are withdrawn at any point:
+```
+[0]:  in = sellAmount * (FEE_DENOMINATOR + feeNumerator) / FEE_DENOMINATOR
+[11]: out_settle = sellAmount - currentBidSum * sellAmount / buyAmount
+[19]: out_fees <= sellAmount * feeNumerator / FEE_DENOMINATOR
+[21]: out_all_order <= sum(out_per_order) <= currentBidSum * sellAmount / buyAmount
+ =>   out <= sellAmount + sellAmount * feeNumerator / FEE_DENOMINATOR
```
Comment (josojo) [inline rounding.md:116]: Nice!

**Review by josojo — APPROVED** [review]
So clean. like it a lot.

**Comment by josojo** [comment]
I like the "linking" system. I think we should merge it today, such that even when we change the explainer, we will not touch the EasyAuction contract anymore.

File: rounding.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+[17]: out_auction_token_per_order = sellAmount * priceNumerator / priceDenominator
+```
+We argue that `sum(out_per_order) <= currentBidSum * sellAmount / buyAmount`.
+Since we must be in case [16] as discussed before, it must be that all orders
+have been summed up [18] and `currentBidSum <= minAuctionedBuyAmount` [16].
+Then:
+```
+ -    currentBidSum = sum(orderSellAmount) <= minAuctionedBuyAmount
+ =>   sum(out_per_order) =  sum(orderSellAmount * sellAmount / buyAmount)
+                         <= sum(orderSellAmount) * sellAmount / buyAmount =  currentBidSum * sellAmount / buyAmount
```
Comment (josojo) [inline rounding.md:81]: And hence, we can conclude in the next line:
```
sum(out_per_order) =  sum(orderSellAmount * sellAmount / buyAmount)
                         <= sum(orderSellAmount) * sellAmount / buyAmount =  currentBidSum * sellAmount / buyAmount
                         <= minAuctionedBuyAmount* sellAmount / buyAmount
                         <= sellAmount
```

**Comment by josojo** [comment]
I am merging it just to make the handover to the audit, such that the EasyAuction.sol does not need to be changed later.

File: rounding.md (hunk 2 above)
Comment (fedgiac) [reply rounding.md:101]: In the expression below the value of `auctioningTokenAmount` is replaced, and I wanted to give a reference to where this replacement comes from.

File: rounding.md (hunk 3 above)
Comment (fedgiac) [reply rounding.md:84]: I tried to be as close as possible to the notation in the code, but these values have different names in different functions.
I tried to use first the one as it appears in the code, and then replace it with naming to make all expressions consistent. I'm not sure how to improve on it without writing down each replacement step.

**Pull request merged**